Repurchase

ValentineReese opened the pull request.

ValentineReese committed:
Add  initialization repurchese code

ValentineReese committed:
Complete repurchese code

ValentineReese committed:
add some unittest for “TimelyReleasePool”

ValentineReese committed:
Completed tests for repurchase

ADLIRUST reviewed (approved):
用 `mpm release last`  --ub 跑测试

element2050 commented on sources/gov/Repurchase.move:
```diff
@@ -0,0 +1,160 @@
+address SwapAdmin {
+
+module Repurchease {
```
代码里回购统一用buyback，对外的文章/公告也统一用的buyback

ValentineReese replied:
好的，名字我改下

element2050 commented on sources/gov/Repurchase.move:
```diff
@@ -0,0 +1,160 @@
+        release_per_time: u128
+    ) acquires EventStore {
+        let sender_address = Signer::address_of(sender);
+        assert!(sender_address == @RepurchaseAccount, Errors::invalid_state(ERROR_NO_PERMISSION));
```
回购的grant资金目前在swap admin账号下，在该处的accept之前，是先转账到buybackAccount？还是sender支持不同的账号？你选择一个

ValentineReese replied:
1. 这里的Repurchase是适用于任意一对Token回购的。
2. 发起者将回购资金质押在合约中

element2050 commented on sources/gov/Repurchase.move:
```diff
@@ -0,0 +1,160 @@
+        let cap = borrow_global<RepurchaseCap<PoolT, ToTokenT>>(broker);
+        let to_token = TimelyReleasePool::withdraw(broker, &cap.cap);
+        let to_token_val = Token::value<ToTokenT>(&to_token);
+        let y_out = TokenSwapRouter::compute_y_out<ToTokenT, FromTokenT>(
```
参数里的fromtoken和to token，参数里调整个顺序感觉会更合理

ValentineReese replied:
这里无论From和To无论谁在前谁在后都会有歧义，换个命名可能比较合理。例如从用户视角来看，BuyTokenT 为STAR，SellTokenT为STC，而从项目方视角来看，则是反过来的。这里回购，以项目方视角比较合理，后面会按照这个参数进行命名修改

element2050 commented on sources/gov/TimelyReleasePool.move:
```diff
@@ -0,0 +1,96 @@
+        assert!(time_interval >= pool.interval, Errors::invalid_state(ERROR_LINEAR_NOT_READY_YET));
+        let times = time_interval / pool.interval;
+
+        let token = Token::withdraw(&mut pool.treasury, (times as u128) * pool.release_per_time);
```
(times as u128) * pool.release_per_time 不对吧，应该是(times as u128) * time_interval  * pool.release_per_time ?

ADLIRUST replied:
> (times as u128) * pool.release_per_time 不对吧，应该是(times as u128) * time_interval * pool.release_per_time ?

`release_per_time`这个在代码逻辑里应该是周期释放量？

ValentineReese replied:
这里是按照周期来释放的，
`  let times = time_interval / pool.interval;` 表示总间隔时间/单次间隔时间=释放次数。
当前释放总额=单次释放额度*释放次数

ValentineReese committed:
Rename for PR comment

ValentineReese committed:
modify for PR comment

element2050 commented on sources/gov/BuyBack.move:
```diff
@@ -0,0 +1,162 @@
+        /// event stream for withdraw
+        accept_event_handle: Event::EventHandle<AcceptEvent>,
+        /// event stream for deposit
+        purchease_event_handle: Event::EventHandle<BuyBackEvent>,
```
buyback没改全

ValentineReese replied:
改好了

ValentineReese committed:
fix rename

ValentineReese committed:
Merge branch 'master' into repurchase

ValentineReese committed:
1. Fix some contract reference error that link XUSDT in bridge
2. Add some return parameter in query_info

ValentineReese committed:
Some testing result

ValentineReese committed:
Change Bridge address to barnard address.

ValentineReese committed:
Fix some error

ValentineReese committed:
1. Use the new event util module named `EventUitl` to refactor the event section of the `Buyback` module
2. Add some functions such as `Deposit`, `Unintialize` to complete the overall process.
3. Add some integration test for Buyback module

ValentineReese committed:
Spectest result.

ValentineReese committed:
Fix the error caused by the wrong parameter input

ValentineReese committed:
When the buyback is executed, the STC released by the pool is actually to exchange STAR in swap pool

ValentineReese committed:
Update dependency

ValentineReese committed:
add some check

ValentineReese committed:
Fix error in BuyBackSTAR::query_info

ValentineReese committed:
Solved the problem of not changing the total number after changing the treasury in the BuyBackSTAR::deposit function

element2050 commented on sources/gov/BuyBack.move:
```diff
@@ -0,0 +1,213 @@
+        EventUtil::init_event_with_T<DissmissEvent>(sender);
+    }
+
+    public fun upgrade_event_struct(sender: &signer) acquires EventStore {
```
为啥还没上线的新功能，就有upgrade方法了？

ValentineReese replied:
这里一会去掉，之前部署了一个事件结构，在去加新事件的时候发现原来的结构不能扩展，所以增加了EventUtil.move来扩展

element2050 commented on sources/gov/BuyBack.move:
```diff
@@ -0,0 +1,213 @@
+        // Deposit to trigger account
+        let amount_y_out = TokenSwapRouter::compute_y_out<BuyTokenT, SellTokenT>(amount_x_in);
+
+        Account::deposit<BuyTokenT>(sender_addr, buy_token);
```
如果回购过程先把回购资金存到触发用户的balance下，回购后再withdraw出来再转移到回购账号下，意味着在用户的transaction列表会增加两笔交易。
有没有可能不经过用户的账号？

ValentineReese replied:
如果不经过用户的账号，则有两种方法可选： 1. 将项目方账户托管，以项目方的身份来进行swap，托管之后则没办法修改参数了。2. 修改现有的swap流程，目前时间不够，delay风险比较高

element2050 replied:
需要考虑一种可能出现的情况，由于滑点的存在，用户的账户进出打不平，多了或者少了STAR，想下会不会出现这个问题，账不过用户的账户是比较安全的。

ValentineReese replied:
这里会有这种问题，如果是这样，那就还是得改整体Swap的流程，需要将原来必传Signer入口拆成可用Token作为入口

element2050 replied:
基于单线程模型的话，交易是原子的，在没有bug的前提下或者并行化之前，过一遍用户账户其实能打平。你可以加个testcase验证一把。
不过这种设计和实现在机制上，是相对容易出问题的。
你评估下，这次是直接上后面再优化，还是改一版再上线。

ValentineReese replied:
这里决定第二版再改

ADLIRUST commented on sources/gov/BuyBackSTAR.move:
```diff
@@ -0,0 +1,83 @@
+        BuyBack::accept<BuyBackSTAR, STAR::STAR, STC::STC>(&sender, total_amount, begin_time, interval, release_per_time);
+    }
+
+    public(script) fun uninit(sender: signer) {
```
uninit 有些怪，`cancellation`怎么样或者 `dismiss`

ValentineReese replied:
这里的uninit和init相对应，用了dissmiss需要有一个与之对应的初始化单词

element2050 commented on sources/gov/BuyBack.move:
```diff
@@ -0,0 +1,213 @@
+        TokenSwapRouter::swap_exact_token_for_token<BuyTokenT, SellTokenT>(sender, amount_x_in, amount_y_out);
+
+        // Withdraw SellToken from swap trigger account
+        Account::deposit<SellTokenT>(broker, Account::withdraw<SellTokenT>(sender, amount_y_out));
```
对于老账户，是不是还得判断是否开启auto_accept ？

ValentineReese replied:
Swap流程里面做了，不需要重复做，参考`TokenSwapRouter::swap_pair_token_auto_accept`

ADLIRUST commented on sources/gov/BuyBackSTAR.move:
```diff
@@ -0,0 +1,83 @@
+        BuyBack::deposit<BuyBackSTAR, STC::STC>(@BuyBackAccount, token);
+    }
+
+    public fun query_info(): (u128, u128, u128, u64, u64, u64, u64, u128, u128) {
```
这里临时封装一个 info 结构体会不会好一些？ @element2050

element2050 commented on sources/gov/BuyBack.move:
```diff
@@ -0,0 +1,213 @@
+        // Deposit to trigger account
+        let amount_y_out = TokenSwapRouter::compute_y_out<BuyTokenT, SellTokenT>(amount_x_in);
+
+        Account::deposit<BuyTokenT>(sender_addr, buy_token);
```
那这个deposit呢？是在swap之前。

ValentineReese replied:
这里的流程是 
1. 从线性释放池里取出STC
2. 根据STC算出STAR额度（调用 `TokenSwapRouter::compute_y_out`）
3. 将STC存入用户账户（也就是你上面提到的deposit）
4. 用户账户去Swap
5. 将swap到的STAR从用户账户取出，取出的额度为上面算出来的额度
6. 将取出的STAR存到基金账户中

ValentineReese committed:
1. Change dependency of bridge
2. Remove old event struct

ValentineReese committed:
remove upgrade function

ValentineReese committed:
Some test results
